### DISCUSSION TITLE
Cannot get past 12.5 on "Shelly Plus PlugS" whatsoever

### DISCUSSION THREAD

**Discussion by zeus86** [arendst/Tasmota#21946] [Support request]
First of: by now I spent at least 3-4 hours crawling github issues and discussions as well as the docs and reflashed the device countless times by now. i tried e.g. `tasmota32solo1.bin` and `tasmota32.bin` of 12.5 as well as the safeboot-versions and some others, manually and via OTA.

My PlusPlugS is still on 12.5, because i did not upgraded for a while, but just today (I know, a bit late) i realized, that i cannot upgrade beyond v12.5 (on a regular-sized image). As i mentioned, I read everything relevant to this topic (I think), so to me the problem is clearly the partition layout. 

However, I tried all the suggested ways of recent as well as older versions of partition manager and wizard just are not working for me.  Either Partition Wizard does not start (and therefore has no menu entry) and fails with 
`bry: failed to run compiled code 'io_error' - inconsistent number of builtin objects.` (current build, sha256sum `6e242a38c99641d8780063b16dc2cafbc9925fa3f3d78113816a3f4d7b81c284  Partition_Wizard.tapp `) or the app does not show meaningful things at all (see screenshots). Unfortunately the docs are very limited regadring this. They describe what the new safeboot-layout looks like and that there is a safeboot image, but thats it, no tutorial how to archieve this at all. On top of that, there aren't good screenshots on how the upgrade process is **supposed to look like** at all, so i'm totally stuck here. 

I also resetted the config and just did an autoconf+wifi-join for the device and tried to start over from there. I can flash 13.0.0-safeboot, but you cannot really do anything meaningful from there without any helpful docs. Heck, i did not even find good documentation on how to explicitly flash (and then boot to) the second partition, because i suppose, you have to boot from there to wipe out or change the first one...

my hw and current partition layout: 

![Screenshot_20240811_011921](https://github.com/user-attachments/assets/38f72a21-52ac-4b54-90cf-1e07903caae6)

in the screenshot above i am on partition `app_0` (at least I suppose that's what the asterisk is for). However, regardless if I am on  `app_0` or `app_1`, when I use an older version of `partition_wizard`, that at least works (sha256sum `77dea377f2726c4a715adacc4f4d3209aecc97709a5185358626978e3f93cf99  Partition_Wizard.tapp`), it does not show anything helpful:
![Screenshot_20240811_012346](https://github.com/user-attachments/assets/e9b26c57-2a04-4ec1-adf6-3b2bd33e368a)
(yes, this is  the whole content of the page)

In this window nothing is clickable at all.  
Not so much better does the partition manager look like:
![Screenshot_20240811_012458](https://github.com/user-attachments/assets/051480ab-0f84-4e17-b2c6-ea7c1be20a81)
(yes, this is also the whole content of this page)

The relevant log-messages i've seen regarding this are this one:  
```
BRY: Exception> 'attribute_error' - the 'Partition_info' object has no attribute 'size'
stack traceback:
<unknown source>: in function `page_show_partition`
<unknown source>: in function `page_show_partitions`
<unknown source>: in function `page_part_mgr`
<unknown source>: in function `<lambda>`
```  

The manager "sometimes" says `no need to click, this is already the active partition` or something like that in the tooltip. However, nothing is clickable at all anyways. As far as i have observed, this also shows always app_0 as active, regardless of what the info-tab says. I think this happens when `app_1` is active, and it looks like this:
![Screenshot_20240811_032229](https://github.com/user-attachments/assets/e7632ec2-7dbb-4b9e-a262-fe13c9094c24)
(the screenshot above shows the partition manager, however, the partition wizard definitively shows app_0 as the active one, even if the info tab says otherwise. If so, the errors in the log are slightly different: 
```
BRY: Exception> 'type_error' - unsupported operand type(s) for +: 'int' and 'nil'
stack traceback:
<unknown source>: in function `page_show_spiffs`
<unknown source>: in function `page_show_partitions`
<unknown source>: in function `page_part_mgr`
<unknown source>: in function `<lambda>`
```

So please, any brave soul out there, please enlighten me how the heck to proceed from here, what I have done wrong, and how to avoid this situation at all. I suppose, what I'm seeing is not normal behaviour, otherwise i think i would have already seen issues regarding this. I am really not frustrated easily with such things, but man, this one made my hair grey...

thanks in advance.

**Comment by zeus86** [comment]
PS.: no, i won't rip apart a plastic-welded plug to flash the plug via serial if someone comes up with that. I know that this will most probably work very easy, but no way......

**Comment by Jason2866** [comment]
Use the matching Tools to your version. See https://github.com/arendst/Tasmota/discussions/18983#discussioncomment-6371024
If PartitionWizard does not work you can try to resize the Partition with PartitionManager. If working this will rebuild the somehow not consistent Partition scheme. Do not try to load safeboot manually. This will not work.
If you have already done. Upgrade device twice with tasmota32-solo1 v12.5

**Reply by zeus86** [reply]
the builds in exactly this thread are what i tried. the older ones from https://github.com/arendst/Tasmota/raw/eb3de9c73a23e5559123dd85df70f844aaa46f33/tasmota/berry/modules/Partition_Wizard.tapp : 
* Partition Manager starts, looks like as seen in the screenshots.
* Partition Wizard starts, looks like as seen in the screenshots. 

for the recent version also mentioned in that thread (https://github.com/arendst/Tasmota/blob/development/tasmota/berry/modules/Partition_Wizard.tapp): 
* partition manager: same as above
* partition wizard: error message as described. 

to be clear: **to my knowledge it is currently not possible to get to a state, where i am able to perform changes alltogether!**

However, i will upgrade to tasmota32-solo1 v12.5 as advised, and get back with updates.

**Comment by zeus86** [comment]
No changes as expected. Here's what i did:

* flashed from `http://ota.tasmota.com/tasmota32/release-12.5.0/tasmota32solo1.bin` via OTA twice. 
* Uploaded manager and wizard from `https://github.com/arendst/Tasmota/raw/eb3de9c73a23e5559123dd85df70f844aaa46f33/tasmota/berry/modules/` and restarted.
* Result:
  * wizard:
![Screenshot_20240813_002451](https://github.com/user-attachments/assets/8ad21bdc-4016-4553-9bf5-a0e6df1b729b)
  * manager: 
![Screenshot_20240813_002510](https://github.com/user-attachments/assets/f1bf2348-d1f8-417d-9bd8-31f39a00b097)
  * log: 
![Screenshot_20240813_002536](https://github.com/user-attachments/assets/cc36111e-c32f-446e-9504-bf709fba42eb)

* then i uploaded the recent version of both tapps from `https://github.com/arendst/Tasmota/blob/development/tasmota/berry/modules/` and restarted.
* Result: 
  * manager:
![Screenshot_20240813_002809](https://github.com/user-attachments/assets/6ec2a4f6-7dd9-4267-bc39-5cdf658744a3)
  * wizard:
    * not available
  * log:
![Screenshot_20240813_003342](https://github.com/user-attachments/assets/38ce7666-81b0-4117-b073-2405715e7526)


* System Info:
  * Model: Shelly Plug Plus S EU (black)
![Screenshot_20240813_002844](https://github.com/user-attachments/assets/aae5966d-1d67-43dc-a47e-abfe9f10c340)
![Screenshot_20240813_002852](https://github.com/user-attachments/assets/1f14e292-6fc8-4d68-8f78-804202140ebf)


I hope this clarifies what the problem is. The things you suggest @Jason2866 i already tried.

**Reply by HFarryranz** [reply]
Same here! Shelly Plus Plug S.
I used today this URL https://github.com/arendst/Tasmota/blob/master/tasmota/berry/modules/Partition_Wizard.tapp 
Starting form  12.5. i was able to start migration.

BUT now  I stuck in SAFEBOOT.

1. device;

Shelly Plus Plug S
Tasmota

Program Version	13.0.0(safeboot)single-core
Build Date & Time	2023-06-26T10:42:56
Core/SDK Version	2_0_10/4.4.5.230614

ESP Chip Id	5809908 (ESP32-U4WDH-D rev.3)
Flash Chip Id	0x164020 (DIO)
Flash Size	4096 KB
Program Flash Size	4096 KB
Program Size	789 KB
Free Program Space	42 KB
Free Memory	156.1 KB (frag. 21%)


Partition safeboot*	832 KB (used 94%)
Partition app0	2688 KB (used 69%)
Partition fs	448 KB


2. device stuck in safeboot
 
Shelly Plus Plug S 4 LEDs
Tasmota Plug_1

Program Version	14.1.0 (release-safeboot) single-core
Build Date & Time	2024-06-03T11:53:23
Core/SDK Version	3_0_0/5.1.4.240529

ESP Chip Id	11913844 (ESP32-U4WDH-D v3.1)
Flash Chip Id	0x164020 (DIO)
Flash Size	4096 KB
Program Flash Size	4096 KB
Program Size	812 KB
Free Program Space	19 KB
Free Memory	212.0 KB (frag. 39%)
Partition safeboot*	832 KB (used 97%)
Partition app0	2688 KB (used 29%)
Partition fs	512 KB

**Reply by Jason2866** [reply]
@HFarryranz Just can assume things happened since logs are hard to get. Probably the OTA convert process was not fully successful or not followed exactly. If the part where the Shelly bootloader is replaced was not successful the symptoms you see does fit.
Trying to update with the Shelly bootloader still there ends in a loop where always the firmware is started at address 0x10000. This is in Tasmota use case the safeboot firmware.
To recover the device only serial flashing is the way.

**Comment by zeus86** [comment]
Does really noone have an idea how to get around my problem?

**Comment by joba-1** [comment]
well my idea is as follows:
Jason2866 says: "To recover the device only serial flashing is the way."
you say: "no way......"
So before you say "I throw it away"
I say: "send it my way," 
I could use one, would be better for the environment anyways ;)

**Reply by zeus86** [reply]
Jason's statement refers to @HFarryranz issue, which is an entirely different issue, because it was soft-bricked half way during an upgrade. On mine i can flash everything up to 12.5, and it works normally, I just cannot get past 12.5 due to the crooked partition layout. 

However,  I don't mind the price of the device, but sending it via mail internationally would most likely be more expensive, than the device itself, so it really isn't environmental friendly. However, this a whole different discussion. In the first place i'd rather would be interested into a solution, because i did nothing special to be in this situation, so i assume, others might fall into this trap also. And no, serial flashing is not really a viable option for plastic-welded devices, at least not for regular usecases.

I really wish there would be more interest in actually finding out what the underlying problem is and how to solve this, in order to prevent such situations in the future.

**Comment by zeus86** [comment]
this discussion can be closed. 

even tough the problem is not solved, the device no longer exists. i was hoping that these things can be dismanteled halfway destruction-free, but no way. they are glued (the glue can be dealt with  some bending indeed), and even if you manage to dismantle the plug without ripping the wifi-foil-antenna apart, there is a riveted PE-connection from top to bottom. beside that, there are no explicitly exposed debug-headers to flash (that in itself wouldn't be a problem, you can attach wires to the PCB-module, but it would be quite fiddly).

so all in all, i destroyed the device during this exercise (somehow a coil inside the device broke off for no reason and was rattling in the device way before i managed to get inside, so it wasn't safe to use anymore anyways...).
